rust-src/id/src/ffi.rs

```diff
@@ -292,6 +298,208 @@ pub extern "C" fn ip_info_cdi_verify_key(
     ptr
 }
 
+/// Create an ArInfo instance from byte array pointers.
+/// Assumes that public_key_ptr points to a PublicKey<G1>
+/// instance and that name_ptr, url_ptr and desc_ptr
+/// to valid utf8 strings, and panics otherwise.
+#[no_mangle]
+unsafe extern "C" fn ar_info_create(
+    identity: u32,
+    public_key_ptr: *mut u8,
+    public_key_len: size_t,
+    name_ptr: *const u8,
+    name_len: size_t,
+    url_ptr: *const u8,
+    url_len: size_t,
+    desc_ptr: *const u8,
+    desc_len: size_t,
+    output_len: *mut size_t,
+) -> *mut ArInfo<G1> {
+    // Identity.
+    let ar_identity = ArIdentity::try_from(identity)
+        .expect("Precondition failed: Unable to create ArIdentity instance from u32.");
+
+    // Public key.
+    let ar_public_key = {
+        let buf = &mut slice_from_c_bytes!(public_key_ptr, public_key_len as usize);
+
+        from_bytes::<elgamal::PublicKey<G1>, &[u8]>(buf).expect(
+            "Precondition failed: Unable to create PublicKey<G1> instance from byte array at \
+             public_key_ptr.",
+        )
+    };
+
+    // Description.
+    let ar_description = {
+        let name = from_utf8(slice_from_c_bytes!(name_ptr, name_len as usize))
+            .expect("Precondition failed: Unable to decode byte array at name_ptr as utf8.")
+            .to_string();
+
+        let url = from_utf8(slice_from_c_bytes!(url_ptr, url_len as usize))
+            .expect("Precondition failed: Unable to decode byte array at url_ptr as utf8.")
+            .to_string();
+
+        let description = from_utf8(slice_from_c_bytes!(desc_ptr, desc_len as usize))
+            .expect("Precondition failed: Unable to decode byte array at desc_ptr as utf8.")
+            .to_string();
+
+        Description {
+            name,
+            url,
+            description,
+        }
+    };
+
+    let ar_info = ArInfo {
+        ar_identity,
+        ar_description,
+        ar_public_key,
+    };
+
+    let bytes = to_bytes(&ar_info);
+    *output_len = to_bytes(&ar_info).len() as size_t;
+    std::mem::forget(bytes);
+    Box::into_raw(Box::new(ar_info))
+}
+
+/// Create a GlobalContext instance from byte array pointers.
```

Comment:
```suggestion
/// Create a [`GlobalContext`] instance from byte array pointers.
```

Reply:
This will make the docs include a link to the GlobalContext type. Please add links when relevant 😊